rfcs/2021-02-23-6531-performance-testing.md

```diff
@@ -0,0 +1,490 @@
+O'Sullivan's criterion does a fair bit more math than Rust's criterion to make
+the results stable, so there is some area for improvement in Rust criterion but
+the fundamental problem remains. It is possible that criterion will gradually
+incorporate instruction count benchmarking, a reasonable proxy for performance,
+but that is very early days yet.
+
+We can see from [Motivating Examples](#motivating-examples) that this benchmark
+work is paying off and has acted as a backstop, disallowing serious regressions
+from making it into releases. A good deal of work has been done to make results
+stable in our Github Actions based CI. PR build times are, however, increasing
+at a steady clip, a drag on productivity and potentially a barrier to
+open-source contributors whose engagement with a change may not be as high as
+full-time employees on the Vector project.
+
+We are actively expanding our use of criterion in the project, as of this
+writing.
+
+### test-harness
+
+The [vector-test-harness](https://github.com/timberio/vector-test-harness/) is a
+"black-box" performance and correctness testing approach. Performance tests
+serve two roles: indicating whether Vector has suffered regressions for given
+workloads and comparing Vector to competitor products. The later role feeds our
+product documentation.  Let's consider the [disk
+buffer](https://github.com/timberio/vector-test-harness/tree/master/cases/disk_buffer_performance)
+performance test. This test is meant to probe the performance characteristic of
+the "disk buffer", the disk backed variant of Vector's
+[`buffers`](https://github.com/timberio/vector/blob/2ac861e09f99036145749ee8af7a7e0d7aa945c6/src/buffers/mod.rs). The
+[README](https://github.com/timberio/vector-test-harness/tree/master/cases/disk_buffer_performance)
+for the test describes the high-level approach: generate as much data as
+possible in 60 seconds and observe the results of the test on average IO
+throughput, CPU consumption and so forth. The test-harness uses
+[ansible](https://github.com/timberio/vector-test-harness/tree/master/cases/disk_buffer_performance/ansible)
+to set up and execute every variant to be tested. System operation during the
+execution is recorded with [dstat](http://dag.wiee.rs/home-made/dstat/) and only
+one iteration of each test is made. Tests are made on AWS spot instance c5.large
+variant. The observations made by dstat are shipped to s3 for long-term storage.
+
+The relatively short duration of this test and its singular instance will tend
+to make the results quite noisy. Spot instances suffer no performance penalty
+compared to reserved capacity but the c5.large may not be representative of the
+machines Vector is deployed to. Only a single build of Vector is run here as
+well. Consider that the
+[0.11.1](https://github.com/timberio/vector/releases/tag/v0.11.1) supports
+aarch64, armv7, x86_64, amd64, arm64, armhf, etc plus a cross product on some of
```

Comment:
D'oh. This is what happens when you use a program to write copy. I will fix the `amd64` thing. What differences exist are not important enough to make the distinction. 

> I'm wondering if we'll want to test the full cross-product or maybe do tiers of support similar to Rust where we focus on x86/arm64 x gnu/musl x Linux first since those are likely to be the most common.

I like the tier model, which is maybe worth calling out in the RFC. At least at first we're _not_ going to have more than x86 being tested and probably only gLibc to start, too.